boards: shields: add new ci test shield

FRASTM opened the pull request:
This PR is introducing a new shield to test various features like:
-  Digital or Analog I/O,
- PWM and UART loopback,
- SD card on SPI bus
- EEPROM or Temp Sensor on I2C.
    
Especially suitable for testing the SDHC card interface over SPI, with the fat_fs application:
` $ west build -b nucleo_f411re samples/subsys/fs/fat_fs -DSHIELD="ci_shield"`

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
_Put a FAT-formatted microSD card in the slot and run test application:_

*** Booting Zephyr OS build zephyr-v2.3.0-886-g73ba6c255a7a  ***                                      
[00:00:00.005,000] <inf> spi_ll_stm32: CS control inhibited (no GPIO device)                          
Sector size 512                                                                                       
Memory Size(MB) 1886                                                                                  
Disk mounted.                                                                                         
                                                                                                      
Listing dir /SD: ...                                                                                  
[DIR ] SYSTEM~1                                                                                       
[FILE] TEST.LOG (size = 6144000)                                                                      
[FILE] TEST_B.LOG (size = 2048000)                                                                    
[00:00:00.049,000] <inf> sdhc_spi: Found a ~1886 MiB SDHC card.                                       
[00:00:00.050,000] <inf> sdhc_spi: Manufacturer ID=3 OEM='SD' Name='SU02G' Revision=0x80 Serial=0x6ab6
[00:00:00.050,000] <inf> main: Block count 3862528                                                    
[00:00:00.098,000] <inf> sdhc_spi: Found a ~1886 MiB SDHC card.                                       
[00:00:00.099,000] <inf> sdhc_spi: Manufacturer ID=3 OEM='SD' Name='SU02G' Revision=0x80 Serial=0x6ab6
[00:00:00.099,000] <inf> spi_ll_stm32: CS control inhibited (no GPIO device)

erwango commented on boards/shields/ci_shield/boards/nucleo_f411re.defconfig:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2020 STMicroelectronics
+#
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_NUCLEO_F411RE
+
+if SPI
+
+config SPI_1
+	default y
+
+endif # SPI
+
+endif # BOARD_NUCLEO_F411RE
```
This file is not needed

erwango commented on boards/shields/ci_shield/boards/nucleo_f411re.overlay:
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2020 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&spi1 {
+        status = "okay";
+        cs-gpios = <&gpiob 6 GPIO_ACTIVE_HIGH>;
+
+        sdhc0: sdhc@0 {
+                status = "okay";
+        };
+};
```
This file should not be needed

FRASTM replied:
ok

erwango commented on boards/shields/ci_shield/Kconfig.defconfig:
```diff
@@ -0,0 +1,15 @@
+
+if SHIELD_CI_SHIELD
+
+config SPI
+	default y
+
+config DISK_ACCESS_SPI_SDHC
+	default y
+
+config DISK_ACCESS_SDHC
+	default y
```
We want application to control activation, so it should be:
```
if DISK_ACCESS_SDHC

config SPI
	default y

config DISK_ACCESS_SPI_SDHC
	default y

endif
```

FRASTM replied:
The application (prj.conf) is configuring CONFIG_DISK_ACCESS=y
So could the `boards/shields/ci_shield/Kconfig.defconfig`  configure with depends on DISK_ACCESS

pabigot commented:
Where can somebody get [one of these](https://os.mbed.com/components/CI-Test-Shield-Minewtech/)?  It's not referenced anywhere on Minew's site anymore; https://github.com/ARMmbed/mbed-HDK is outdated; and I can't find anybody via Google that can ship one.

IIRC there is a plan to produce something like this for Zephyr, though maybe that's been abandoned.

jfischer-phytec-iot commented:
> IIRC there is a plan to produce something like this for Zephyr, though maybe that's been abandoned.

Yes, I resumed my work on it.

FRASTM commented:
- removing boards/shields/ci_shield/boards/xxx

- boards/shields/ci_shield/Kconfig.defconfig depending on DISK_ACCESS

erwango commented:
> Where can somebody get one of these? It's not referenced anywhere on Minew's site anymore; https://github.com/ARMmbed/mbed-HDK is outdated; and I can't find anybody via Google that can ship one.

That's a good point. We have a few ones internally. I'm not sure how people can get those now. Since schematics are open source, potentially anyone could make some more (recent version is available here: https://github.com/ARMmbed/mbed-HDK-Eagle-Projects/tree/master/Shields/CITestShield). But, as you mention, this doesn't mean people can get them easily.
Based on this, should we consider it can't go upstream, since not publicly available ?

To my understanding, this would be the exact same issue for a potential Zephyr CI test shield. @jfischer-phytec-iot do you confirm?

These shields have definitely an interest for developers, so despite the fact that they can hardly be distributed for now, i'd see an interest in having them available, maybe as a kind of "bootstrap" to show this kind of thing has interest and some maker start producing and distributing it...

jfischer-phytec-iot commented:
> To my understanding, this would be the exact same issue for a potential Zephyr CI test shield. @jfischer-phytec-iot do you confirm?

Yes, it would offer the same set of functionality.

erwango commented:
> ld offer the same set of functionality.

Sure, but I was rather pointing at the lack of public distribution (at least this was the status the last time we mentioned it).

jfischer-phytec-iot commented:
> > ld offer the same set of functionality.
> 
> Sure, but I was rather pointing at the lack of public distribution (at least this was the status the last time we mentioned it).

They would be distributed the same way as the reel and link boards, means with the same issues :-)

pabigot commented:
> Based on this, should we consider it can't go upstream, since not publicly available ?

That was my point, yes.   We do not have a clear policy regarding the addition or removal of support for hardware that is obsolete, cannot be purchased, or lacks publicly available documentation.  #20548 for example.

galak reviewed (commented):
Just curious if you know if these shields are something one can buy or not?

FRASTM committed:
disk: sdhc: use the multiple block transfer for writing

The sdhc_spi_write_multi function is sending the CMD25
for multiple block write operation on the SD card
instead of looping the CMD24. This is improving
the transfer speed compared to the sdhc_spi_write function
when the nb of blocks to write is significant (more than 2).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: fs: spi_sd test application for write speed evaluation

This application is delivered as example
to measure read/write speed on a SD card.
Just adapt the main function.
It comes from the fat_fs sample and code example from
Tavish Naruka <tavishnaruka@gmail.com>

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented:
> Just curious if you know if these shields are something one can buy or not?

@galak, we got made the ones we have, but there is no distribution channel from what I know.

FRASTM committed:
boards: shields: add new ci_shield for test arduino compatible

Adding this CI_Test Shield to test various features like
Digital or Analog I/O,
PWM and UART loopback,
SD card on SPI bus
EEPROM or Temp Sensor on I2C.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase

gmarull commented:
FYI I've created a Zephyr clone of the ARM mbed test shield, haven't produced the board yet so there's no guarantee that current design will work. https://github.com/teslabs/zephyr-ci-shield cc: @FRASTM @galak @erwango @pabigot
